src/drivers/imx/sdma.c

```diff
@@ -730,47 +728,38 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
 		 * For MEM2MEM the source is stored in buf_addr and destination
 		 * is in buf_xaddr.
 		 */
-		bd->buf_xaddr = 0;
-		bd->buf_addr = config->elem_array.elems[i].src;
-		if (!src_may_change)
-			bd->buf_addr =
-				config->elem_array.elems[i].dest;
-		if (src_may_change && dst_may_change) {
-			/* M2M copy */
-			bd->buf_xaddr =
-				config->elem_array.elems[i].dest;
-		}
-
-		/* We are interested in the element width on the hardware side;
-		 * on RAM side we can adjust.
-		 */
-		width = config->src_width;
-		if (!dst_may_change)
+		switch (config->direction) {
+		case DMA_DIR_MEM_TO_DEV:
+			bd->buf_addr = config->elem_array.elems[i].src;
+			width = config->src_width;
+			break;
+		case DMA_DIR_DEV_TO_MEM:
+			bd->buf_addr = config->elem_array.elems[i].dest;
 			width = config->dest_width;
+			break;
+		case DMA_DIR_MEM_TO_MEM:
+			bd->buf_addr = config->elem_array.elems[i].src;
+			bd->buf_xaddr = config->elem_array.elems[i].dest;
+			width = config->dest_width;
+			break;
+		default:
+			return -EINVAL;
+		}
 
 		bd->config =
 			SDMA_BD_COUNT(config->elem_array.elems[i].size) |
 			SDMA_BD_CMD(SDMA_CMD_XFER_SIZE(width)) | SDMA_BD_CONT;
+
 		if (!config->irq_disabled)
 			bd->config |= SDMA_BD_INT;
-		if (dst_may_change) {
-			/* Capture or M2M, enable this descriptor to be
-			 * used by SDMAC
-			 */
-			bd->config |= SDMA_BD_DONE;
-			/* On playback we don't do this and instead wait
-			 * for copy() to let us know the data is ready.
-			 * copy() is called during preload.
-			 */
-		}
-		if (src_may_change && dst_may_change)
-			bd->config |= SDMA_BD_EXTD;
+
+		bd->config |= SDMA_BD_DONE | SDMA_BD_EXTD |  SDMA_BD_CONT;
```

Comment:
This looks like a behavior change, is setting the EXTD bit the correct thing?

Reply:
Yes, you are right will fix this.

Reply:
We have to put for all because SOF expects an interrupt for each of its elems.